docs/developing_iree/profiling_cpu_events.md

```diff
@@ -1,13 +1,247 @@
+
+On Android we can use
+[`simpleperf`](https://developer.android.com/ndk/guides/simpleperf). It's
+preinstalled on current Android `userdebug` images, and part of the Android NDK.
+
+In theory, as Android is Linux, it should be possible to use `perf`.
+Unfortunately, `perf` is difficult to build for Android. Fortunately,
+`simpleperf` is readily available: it is preinstalled in Android `userdebug`
+images, and it is part of the Android NDK.
+
+First, we record on the device:
+```shell
+adb shell \
+  simpleperf record -e raw-l1d-cache-refill -o /data/local/tmp/perf.data \
+    /data/local/tmp/iree-benchmark-module \
+      --driver=dylib \
+      ... command-line arguments of iree-benchmark-module as usual ...
+```
+
+Then pull the recorded data from the device, and analyze on the desktop. We
+assume that `${ANDROID_NDK}` points to the local copy of the Android NDK.
+```shell
+adb pull /data/local/tmp/perf.data /tmp/perf.data
+${ANDROID_NDK}/simpleperf/report.py -i /tmp/perf.data
+```
+
+This prints a breakdown of `raw-l1d-cache-refill` events by symbol.
+
+Like with `perf`, a list of event types can be queried by the `list` subcommand:
+
+```shell
+adb shell simpleperf list
+```
+
+#### No support for `annotate` by CPU event
+
+There is no `simpleperf annotate`. The `simpleperf` documentation lists a couple
+of
+[ways](https://android.googlesource.com/platform/system/extras/+/master/simpleperf/doc/README.md#show-annotated-source-code-and-disassembly)
+of achieving the same thing.
+
+However:
+* The common case of annotating by time, as opposed to annotating by CPU event,
+  is supported by [Tracy](profiling_with_tracy.md).
+* Annotating by CPU event is inherently not working due to hardware limitations
+  of the ARM CPUs found in Android devices. That is, the hardware is too
+  imprecise at pinning an event to a particular instruction.
+
+## Interpreting CPU event counts
+
+### Problems
+
+There are multiple layers of complexity in interpreting CPU event counts.
+
+#### These events are in themselves normal
+
+The first difficulty is in the fact that most of these events are *normal*. So
+just knowing that they happened is not in itself actionable.
+
+For example, if we learn that some code causes cache misses, that isn't big
+news: so does all code. Maybe this code has *too many* cache misses, but how
+many is too many? Maybe this code alone accounts for a large fraction of the
+overall total of the whole program, but maybe even that is normal, for instance
+if the code being studied is the 'hot' part of the program where a large
+fraction of overall time is spent?
+
+#### These events are hardware-dependent and under-documented
+
+Many of these events have a meaning that varies between CPUs and that is
+difficult to characterize on any CPU, let alone in a way that applies to all
+CPUs.
+
+For example, take the "L2 data cache refill". On ARM, with `simpleperf`, that
+would be `raw-l2d-cache-refill`. Questions:
+* Is “L2” inclusive of “L1”?
```

Comment:
link to https://en.wikipedia.org/wiki/Cache_inclusion_policy

Reply:
done, thanks!